File: docs/src/main/java/jexamples/networking/pingpong/Ponger.java
package jexamples.networking.pingpong;

import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Positive;
import se.sics.kompics.Handler;
import se.sics.kompics.network.Network;

public class Ponger extends ComponentDefinition {

  Positive<Network> net = requires(Network.class);

  private long counter = 0;
  private final TAddress self;

  public Ponger(Init init) {
    this.self = init.self;
  }

  Handler<Ping> pingHandler =
      new Handler<Ping>() {
        public void handle(Ping event) {
          counter++;
          logger.info("Got Ping #{}!", counter);
          trigger(new Pong(self, event.getSource()), net);
        }
      };

  {
    subscribe(pingHandler, net);
  }

  public static class Init extends se.sics.kompics.Init<Ponger> {
    public final TAddress self;

    public Init(TAddress self) {
      this.self = self;
    }
  }
}

File: docs/src/main/scala/sexamples/networking/pingpongdistributed/Ponger.scala
package sexamples.networking.pingpongdistributed

import se.sics.kompics.sl._
import se.sics.kompics.network.Network

class Ponger(init: Init[Ponger]) extends ComponentDefinition {

  val Init(self: TAddress) = init;

  val net = requires[Network];

  private var counter: Long = 0L;

  net uponEvent {
    case ping: Ping => {
      counter += 1L;
      log.info(s"Got Ping #${counter}!");
      trigger(Pong(self, ping.getSource()) -> net);
    }
  }

}
